docs/arch/adr-011-kurl-addon-operator.md

```diff
@@ -0,0 +1,96 @@
+
+## Solution
+
+The operator will be implemented with the following objectives in mind:
+
+- It will be scaffolded using [kubebuilder](https://book.kubebuilder.io/introduction.html).
+- It will be developed using [SOLID](https://en.wikipedia.org/wiki/SOLID) principles and operator best practices.
+- It will distribution agnostic. I.e. it will not be tightly coupled to the kURL Kubernetes distribution.
+- It will manage application add-ons _only_ and _exclude_ managing CRI and CNI resources.
+- It will install and confiure the KOTS admin console.
+- It can be installed on an existing [kURL](https://kurl.sh/) cluster with only Kubernetes, containerd and flannel.
+- The installation mechanism for add-ons will be through the use of Helm Charts.
+- By default the add-on configuration will be opinionated by us but we will enable the end-user to override the configuration for maximum flexibility. Vendors and users will be able to override the configuration using a standard [Helm values](https://helm.sh/docs/chart_template_guide/values_files/) file.
+- It will rely as much as possible on Helm for managing the lifecyle of the add-ons.
+
+### The `Installer` API
+
+The `Installer` API defines a resource for installing specific add-ons.
+
+```yaml
+apiVersion: cluster.kurl.sh/v2beta1
+kind: Installer
+metadata:
+  name: installer-sample
+spec:
+  channel: "1.27"
+  minio:
+    enabled: true
+    configOverride:
+        persistence:
+            enabled: true
+        replicas: 1
+        mode: standalone
+
+```
+
+### The Controllers
+
+The operator will have at minimum one controller which ensures the `Installer` CR is reconciled with the cluster periodically. The `Installer` controller (main controller) will delegate responsiblity for managing add-ons to other controllers. The reconcilation for the main controller will consist of determining the add-ons to be managed from the `Installer` CR and creating a controller for each add-on (addon-on controller).
+
+The add-on controller will be responsible for
+- verifying the requirements for the add-on are met
+- managing the add-on itself
+- ensure add-on is in desired state
+- update the add-on CR status
+
+### Managing the Add-ons
+
+How the add-ons will be managed is still an open question. We have a couple of options:
+
+_Use the Helm Controller provided by Flux_
+
+The [Flux Helm Controller](https://fluxcd.io/flux/components/helm/helmreleases/) affords us the following benefits:
+- Stable [HelmRelease API](https://fluxcd.io/flux/components/helm/api/)
+- Helm release [dependencies](https://fluxcd.io/flux/components/helm/helmreleases/#helmrelease-dependencies)
```

Comment:
It is indeed a cool feature, but it appears to be more suitable for working within tooling environments rather than building an Operator and our use case. I would think that would be very nice if we have only a tooling.  

Consider the trade-off between to create a HelmRelease CR with dependencies versus directly invoking Helm to install the dependencies before deploying the HelmChart.

By using pure Helm we have greater control over the order and state of the dependencies before adding our HelmChart. However, with a HelmController managing the dependencies, we may face uncertainties about its reliability. In the event of a problem with a dependency, it raises concerns about how we can work around such issues. If we attempt to interact with Helm directly to address potential pitfalls, it can introduce complexities such as race conditions that are challenging to handle effectively in any project.

In summary, while the HelmController feature offers convenience within tooling environments, it may not provide the same level of reliability and control as using Helm directly when we are building an Operator. We need to carefully consider the implications, potential workarounds, and complexities that may arise when managing dependencies through a HelmController in our specific project context.

Reply:
This is true, using another tool does mean loss of some control. It also means less effort to implement and maintain going forward. Our decision needs to be based on both and just stating that it's more precise without addressing the implementation and maintenance costs is not addressing the full story. Our dev time is limited, loosing precise control in favor of not having to implement and maintain is one of our goals moving to a new implementation.

My concern here is, why would we want to manage dependencies explicitly that doesn't sound like an advantage to me unless you have examples of bad decisions flux has made doing this that we need to implement in another way.